notebooks/03_evaluate/comparison_movielens.ipynb

```diff
@@ -0,0 +1,982 @@
+    "        total_users = np.array(total_users[1:])\n",
+    "        test_users = test[USER_COL].unique()\n",
+    "        test_users = np.intersect1d(test_users, total_users)\n",
+    "        users_items = cartesian_product(test_users, total_items)\n",
+    "        users_items = pd.DataFrame(users_items, columns=[USER_COL, ITEM_COL])\n",
+    "        training_removed = pd.concat([users_items, train[[USER_COL, ITEM_COL]]]).drop_duplicates(keep=False)\n",
+    "        top_k_scores = score(model, \n",
+    "                             test_df=training_removed,\n",
+    "                             user_col=USER_COL, \n",
+    "                             item_col=ITEM_COL, \n",
+    "                             prediction_col=PREDICTION_COL, \n",
+    "                             top_k=TOP_K)\n",
+    "    return top_k_scores, t\n",
+    "\n",
+    "\n",
+    "def recommend_k_ncf(model, test, train):\n",
+    "    with Timer() as t: \n",
+    "        users, items, preds = [], [], []\n",
+    "        item = list(train[ITEM_COL].unique())\n",
+    "        for user in train[USER_COL].unique():\n",
+    "            user = [user] * len(item) \n",
+    "            users.extend(user)\n",
+    "            items.extend(item)\n",
+    "            preds.extend(list(model.predict(user, item, is_list=True)))\n",
+    "        top_k_scores = pd.DataFrame(data={USER_COL: users, ITEM_COL:items, PREDICTION_COL:preds})\n",
+    "        merged = pd.merge(train, top_k_scores, on=[USER_COL, ITEM_COL], how=\"outer\")\n",
+    "        top_k_scores = merged[merged[RATING_COL].isnull()].drop(RATING_COL, axis=1)\n",
+    "    return top_k_scores, t\n",
+    "\n",
+    "\n",
+    "ranking_predictor = {\n",
+    "    \"als\": lambda model, test, train: recommend_k_als(model, test, train),\n",
+    "    \"sar_single_node\": lambda model, test, train: recommend_k_sar_single_node(model, test, train),\n",
+    "    \"svd\": lambda model, test, train: recommend_k_svd(model, test, train),\n",
+    "    \"fastai\": lambda model, test, train: recommend_k_fastai(model, test, train),\n",
+    "    \"ncf\": lambda model, test, train: recommend_k_ncf(model, test, train),\n",
+    "}"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "Metrics"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 13,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "def rating_metrics_pyspark(test, predictions):\n",
+    "    rating_eval = SparkRatingEvaluation(test, \n",
+    "                                        predictions, \n",
+    "                                        col_user=USER_COL, \n",
+    "                                        col_item=ITEM_COL, \n",
+    "                                        col_rating=RATING_COL, \n",
+    "                                        col_prediction=PREDICTION_COL)\n",
+    "    return {\n",
+    "        \"RMSE\": rating_eval.rmse(),\n",
```

Comment:
I think you could also rewrite this as a for loop, like for the python metrics below. Here you need to use `getattr()`. It is a bit less explicit though.